examples/multi_gpu/papers100m_gcn_multinode.py

```diff
@@ -54,97 +41,128 @@ def run(world_size, data, split_idx, model):
+        torch.cuda.synchronize()
+        num_batches = i + 1.0
         if rank == 0:
-            sec_per_iter = (time.time() - start) / (i - warmup_steps)
+            sec_per_iter = (time.time() - start) / (num_batches - warmup_steps)
             print(f"Avg Training Iteration Time: {sec_per_iter:.6f} s/iter")
-
-            model.eval()
-            total_correct = total_examples = 0
-            for i, batch in enumerate(val_loader):
-                if i >= val_steps:
-                    break
-                if i == warmup_steps:
-                    start = time.time()
-
-                batch = batch.to(device)
-                with torch.no_grad():
-                    out = model(batch.x, batch.edge_index)[:batch.batch_size]
-                pred = out.argmax(dim=-1)
-                y = batch.y[:batch.batch_size].view(-1).to(torch.long)
-
-                total_correct += int((pred == y).sum())
-                total_examples += y.size(0)
-
-            print(f"Val Acc: {total_correct / total_examples:.4f}")
-            sec_per_iter = (time.time() - start) / (i - warmup_steps)
-            print(f"Avg Inference Iteration Time: {sec_per_iter:.6f} s/iter")
-
-    if rank == 0:
         model.eval()
-        total_correct = total_examples = 0
-        for i, batch in enumerate(test_loader):
+        acc_sum = 0.0
+        for i, batch in enumerate(val_loader):
+            if i >= val_steps:
+                break
+            if i == warmup_steps:
+                torch.cuda.synchronize()
+                start = time.time()
+
             batch = batch.to(device)
+            batch_size = batch.batch_size
             with torch.no_grad():
-                out = model(batch.x, batch.edge_index)[:batch.batch_size]
-            pred = out.argmax(dim=-1)
-            y = batch.y[:batch.batch_size].view(-1).to(torch.long)
-
-            total_correct += int((pred == y).sum())
-            total_examples += y.size(0)
-        print(f"Test Acc: {total_correct / total_examples:.4f}")
+                out = model(batch.x, batch.edge_index)[:batch_size]
+            acc_sum += acc(out[:batch_size].softmax(dim=-1),
+                           batch.y[:batch_size])
+        acc_sum = torch.tensor(float(acc_sum), dtype=torch.float32,
+                               device=device)
+        dist.all_reduce(acc_sum, op=dist.ReduceOp.SUM)
+        num_batches = torch.tensor(float(i + 1), dtype=torch.float32,
+                                   device=acc_sum.device)
+        dist.all_reduce(num_batches, op=dist.ReduceOp.SUM)
+        torch.cuda.synchronize()
+        if rank == 0:
+            print(
+                f"Validation Accuracy: {acc_sum/(num_batches) * 100.0:.4f}%", )
+            sec_per_iter = (time.time() - start) / (num_batches - warmup_steps)
+            print(f"Avg Inference Iteration Time: {sec_per_iter:.6f} s/iter")
+    dist.barrier()
+
+    model.eval()
+    acc_sum = 0.0
+    for i, batch in enumerate(test_loader):
```

Comment:
Can't we out-source to its own function to avoid the duplication between val/test?

Reply:
yes